src/box/applier.cc

```diff
@@ -1252,6 +1341,563 @@ applier_on_rollback(struct trigger *trigger, void *event)
+			if (msg != NULL)
+				break;
+			fiber_yield();
+			if (fiber_is_cancelled())
+				return 0;
+		} while (true);
+		applier_thread_push_tx(thread, msg, tx);
+	}
+	return 0;
+exit_notify:
+	/* Notify the tx thread that its applier exited with an error. */
+	assert(!diag_is_empty(diag_get()));
+	diag_move(diag_get(), &applier->thread.exit_msg.diag);
+	cpipe_push(&thread->tx_pipe, &applier->thread.exit_msg.base.base);
+	return 0;
+}
+
+/** The main applier thread fiber function. */
+static int
+applier_thread_f(va_list ap)
+{
+	struct applier_thread *thread = va_arg(ap, typeof(thread));
+	int rc = cbus_endpoint_create(&thread->endpoint, cord()->name,
+				      fiber_schedule_cb, fiber());
+	assert(rc == 0);
+	(void)rc;
+
+	cpipe_create(&thread->tx_pipe, "tx_prio");
+
+	cbus_loop(&thread->endpoint);
+
+	unreachable();
+}
+
+/** Initialize and start the applier thread. */
+static void
+applier_thread_create(struct applier_thread *thread)
+{
+	static int thread_id = 0;
+	const char *name = tt_sprintf("applier_%d", ++thread_id);
+
+	memset(thread, 0, sizeof(*thread));
+
+	if (cord_costart(&thread->cord, name, applier_thread_f, thread) != 0)
+		diag_raise();
+
+	cpipe_create(&thread->thread_pipe, name);
+}
+
+/** Alive applier thread list. */
+static struct applier_thread **applier_threads;
+
+/**
+ * A pointer to the thread which will accept appliers once thread count reaches
+ * the maximum configured value.
+ */
+static int fill_thread = 0;
+
+void
+applier_init(void)
+{
+	assert(replication_threads > 0);
+	applier_threads =
+		(struct applier_thread **)xmalloc(replication_threads *
+						  sizeof(struct applier_thread *));
+	for (int i = 0; i < replication_threads; i++) {
+		struct applier_thread *thread =
+			(struct applier_thread *)xmalloc(sizeof(*thread));
+		applier_thread_create(thread);
+		applier_threads[i] = thread;
+	}
+}
+
+void
+applier_free(void)
+{
+	for (int i = 0; i < replication_threads; i++) {
+		struct applier_thread *thread = applier_threads[i];
+		tt_pthread_cancel(thread->cord.id);
+		tt_pthread_join(thread->cord.id, NULL);
+		free(thread);
+	}
+	free(applier_threads);
+}
+
+/** Get a working applier thread. */
+static struct applier_thread *
+applier_thread_next(void)
+{
+	struct applier_thread *thread = applier_threads[fill_thread];
+	fill_thread = (fill_thread + 1) % replication_threads;
+	return thread;
+}
+
+static void
+applier_msg_init(struct applier_msg *msg, struct applier *applier, cmsg_f f)
+{
+	cmsg_init(&msg->base, tx_route);
+	msg->applier = applier;
+	msg->f = f;
+}
+
+/** Initialize the ibuf used by applier thread. */
+static void
+applier_thread_ibuf_init(struct applier *applier)
+{
+	ibuf_create(&applier->thread.ibuf, &cord()->slabc, 1024);
+	/*
+	 * Move unparsed data, if any, from the previously used tx ibuf to the
+	 * new buf.
+	 */
+	size_t parse_size = ibuf_used(&applier->ibuf);
+	if (parse_size > 0)
+		ibuf_move_tail(&applier->ibuf, &applier->thread.ibuf, parse_size);
```

Comment:
If we could use the applier thread for join, we wouldn't need to maintain two buffers and could also get rid of the virtualization layer. Not sure it's possible though.

Reply:
Final join can be transferred to thread quite easily, but I don't like the idea that some part of JOIN would happen in thread, while the other part would still live in tx.

Initial join could be transferred as well, but not so easily.